### ISSUE TITLE
EKS-Anywhere Stop Working exiting all running Docker Containers

### ISSUE THREAD

**Issue by minasys** [aws/eks-anywhere#684]
**What happened**:
I am running EKS-A using Docker provider. it goes successfully building it, and I can connect to it and list nodes and namespaces for testing. but after an hour or so if I run the kubectl get nodes it cannot find the nodes anymore, to find taht all the docker instances are exited. giving exit (1) and (137) codes. the -lb container shows the following log lines:

```
 Admin-OneClick:~/environment $ docker logs gitops-lb
[WARNING] 322/055153 (1) : config : missing timeouts for frontend 'controlPlane'.
   | While not properly invalid, you will certainly encounter various problems
   | with such a configuration. To fix this, please ensure that all following
   | timeouts are set to a non-zero value: 'client', 'connect', 'server'.
[NOTICE] 322/055153 (1) : New worker #1 (8) forked
[WARNING] 322/055223 (1) : Reexecuting Master process
[NOTICE] 322/055223 (1) : haproxy version is 2.2.9-2~bpo10+1
[NOTICE] 322/055223 (1) : path to executable is /usr/sbin/haproxy
[ALERT] 322/055223 (1) : sendmsg()/writev() failed in logger #1: No such file or directory (errno=2)
[WARNING] 322/055223 (8) : Stopping frontend controlPlane in 0 ms.
[WARNING] 322/055223 (8) : Stopping backend kube-apiservers in 0 ms.
[WARNING] 322/055223 (8) : Stopping frontend GLOBAL in 0 ms.
[WARNING] 322/055223 (8) : Proxy controlPlane stopped (cumulated conns: FE: 21, BE: 0).
[WARNING] 322/055223 (8) : Proxy kube-apiservers stopped (cumulated conns: FE: 0, BE: 21).
[WARNING] 322/055223 (8) : Proxy GLOBAL stopped (cumulated conns: FE: 0, BE: 0).
[NOTICE] 322/055223 (1) : New worker #1 (24) forked
[WARNING] 322/055223 (24) : Server kube-apiservers/gitops-kb5fp is DOWN, reason: Layer4 connection problem, info: "SSL handshake failure", check duration: 0ms. 0 active and 0 backup servers left. 0 sessions active, 0 requeued, 0 remaining in queue.
[ALERT] 322/055223 (24) : backend 'kube-apiservers' has no server available!
[WARNING] 322/055223 (1) : Former worker #1 (8) exited with code 0 (Exit)
[WARNING] 322/055258 (24) : Server kube-apiservers/gitops-kb5fp is UP, reason: Layer7 check passed, code: 200, check duration: 4ms. 1 active and 0 backup servers online. 0 sessions requeued, 0 total in queue.
[WARNING] 322/071801 (1) : Exiting Master process...
[WARNING] 322/071801 (24) : Stopping frontend control-plane in 0 ms.
[WARNING] 322/071801 (24) : Stopping backend kube-apiservers in 0 ms.
[WARNING] 322/071801 (24) : Stopping frontend GLOBAL in 0 ms.
[WARNING] 322/071801 (24) : Proxy control-plane stopped (cumulated conns: FE: 262, BE: 0).
[WARNING] 322/071801 (24) : Proxy kube-apiservers stopped (cumulated conns: FE: 0, BE: 262).
[WARNING] 322/071801 (24) : Proxy GLOBAL stopped (cumulated conns: FE: 0, BE: 0).
```

I increased the cloud9 instance from t3.large (8 GB) to t3.xlarge (16 GB) of memory

**What you expected to happen**:
to have the cluster continue running, and to stop it when I want to stop it.

**How to reproduce it (as minimally and precisely as possible)**:
create the EKS-A cluster based on Docker provider with Gitops enabled and the count for both the worker nodes and controleplane nodes 2. and etcd just 1

wait an hour or so leaving it. will come back to find it exited and if you run kubectl command it will not work for you.

**Anything else we need to know?**:
I am running it on a Cloud9 instance (t3.xlarge).

**Environment**:
- EKS Anywhere Release: v0.5.0
- EKS Distro Release: 1.21

**Comment by pokearu** [comment]
Hi @minasys thanks for reaching out.

This does seem like a memory issue with docker. We use KinD for our local docker clusters and we do recommend allocating 6GB of memory to docker.
It could be that the docker runtime ran out of memory and the containers were killed.

Could you share more info regarding the memory allocations?

**Comment by minasys** [comment]
I am not setting any memory or CPU limits on the running  containers. Here is the output of teh docker stats command for the running eks-a containers
```
CONTAINER ID   NAME                               CPU %     MEM USAGE / LIMIT     MEM %     NET I/O           BLOCK I/O         PIDS
8c1011565e0f   dev-cluster-md-0-d9995bdfb-chndf   1.27%     251.5MiB / 31.07GiB   0.79%     155MB / 1.07MB    193kB / 82.9kB    92
2af607c6d667   dev-cluster-md-0-d9995bdfb-r8db6   5.68%     650.8MiB / 31.07GiB   2.05%     538MB / 17.9MB    451kB / 82.4kB    545
b2efba44d7bc   dev-cluster-gjnj9                  8.27%     1.268GiB / 31.07GiB   4.08%     554MB / 144MB     2.18MB / 2.49MB   492
9d292beb6555   dev-cluster-etcd-jck8t             3.12%     135.4MiB / 31.07GiB   0.43%     45.9MB / 117MB    39.4MB / 83.5kB   32
961361b32501   dev-cluster-lb                     0.02%     17.02MiB / 31.07GiB   0.05%     68.1MB / 67.3MB   12MB / 0B         9
^C
Admin-FullAccess:~/environment $ 
```
I also redeployed the cluster, disabling GitOps, and just running 1 CP, 1 Etcd, and 2 WN, on a t3.2xlarge (32 GB of Memory) but I am still facing same issue.

**Comment by mdsgabriel** [comment]
Thank you for using EKS Anywhere.

I believe this might be related to the restart policy used by the containers. 
The containers created by the Docker provider have the[ default restart policy](https://docs.docker.com/config/containers/start-containers-automatically/) and will not be restarted if you restart Docker.

You can verify this by running the following command:

```
docker inspect <container id> | grep -i policy -A 3
```

You should see something like 
```
   "RestartPolicy": {
                "Name": "no",
                "MaximumRetryCount": 0
            },
```

You can change the restart policy by running 

```
docker update <container id> --restart unless-stopped
```

But please have in mind that the Docker provider is meant for quick demos and not for long running servers.

**Comment by jaxesn** [comment]
This will be fixed in our next (0.7.0) release.
